src/treemodel/composer/modifyselection.js

```diff
@@ -27,20 +27,23 @@ export default function modifySelection( selection, options = {} ) {
 	const focus = selection.focus;
 	const walker = new TreeWalker( {
 		boundaries: getSearchRange( focus, isForward ),
-		singleCharacters: true
+		singleCharacters: true,
+		direction: isForward ? 'FORWARD' : 'BACKWARD'
 	} );
 
-	const items = Array.from( walker );
-	let next = items[ isForward ? 'shift' : 'pop' ]();
+	let next = walker.next();
 
 	// 1. Nothing to do here.
-	if ( !next ) {
+	if ( next.done ) {
 		return;
 	}
 
+	// Replace TreeWalker step wrapper by clean step value.
+	next = next.value;
```

Comment:
I just realised that our code is, in fact, static typed. And, even more spuriously realised, that it is what I like in it. `next` and `value` are different types and overwriting one with another feel like a bad small, because of this static typing.